### PR TITLE
Setup integration tests

### PR DESCRIPTION
**Pull request by suevalov**

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "extends": "standard",
+  "extends": ["standard", "plugin:prettier/recommended"],
```
Comment (jelz) [inline .eslintrc.json]: It would be nice to introduce prettier in a separate PR.

File: test-extensions/test-field-extension/build/index.html [hunk 2]
```diff
@@ -0,0 +1 @@
...nstable_now()};var R=new MessageChannel,F=R.port2;R.port1.onmessage=function(){O=!1;var e=M,t=L;M=null,L=-1;var r=n.unstable_now(),a=!1;if(0>=H-r){if(!(-1!==t&&t<=r))return z||(z=!0,E(V)),M=e,void(L=t);a=!0}if(null!==e){D=!0;try{e(a)}finally{D=!1}}};var V=function(e){if(null!==M){E(V);var t=e-H+B;t<B&&A<B?(8>t&&(t=8),B=t<A?A:t):A=t,H=e+B,O||(O=!0,F.postMessage(void 0))}else z=!1};T=function(e,t){M=e,L=t,D||0>t?F.postMessage(void 0):z||(z=!0,E(V))},x=function(){M=null,O=!1,L=-1}}n.unstable_ImmediatePriority=1,n.unstable_UserBlockingPriority=2,n.unstable_NormalPriority=3,n.unstable_IdlePriority=5,n.unstable_LowPriority=4,n.unstable_runWithPriority=function(e,t){switch(e){case 1:case 2:case 3:case 4:case 5:break;default:e=3}var r=i,a=l;i=e,l=n.unstable_now();try{return t()}finally{i=r,l=a,f()}},n.unstable_next=function(e){switch(i){case 1:case 2:case 3:var t=3;break;default:t=i}var r=i,a=l;i=t,l=n.unstable_now();try{return e()}finally{i=r,l=a,f()}},n.unstable_scheduleCallback=function(e,t){var r=-1!==l?l:n.unstable_now();if("object"==typeof t&&null!==t&&"number"==typeof t.timeout)t=r+t.timeout;else switch(i){case 1:t=r+-1;break;case 2:t=r+250;break;case 5:t=r+1073741823;break;case 4:t
```
Comment (jelz) [inline test-extensions/test-field-extension/build/index.html]: Do we need to commit this?

File: test-extensions/test-field-extension/extension.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (jelz) [inline test-extensions/test-field-extension/extension.json]: can we put them to `test/extensions` instead of new top-level directory?

File: test/cypress/fixtures/example.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (jelz) [inline test/cypress/fixtures/example.json]: can be deleted?

File: test/cypress/integration/field-extension.spec.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  it('opens first post and checks that field extension is rendered', () => {
+    cy.visit(entry(post.id))
+
+    cy.getByText(post.title).should('exist')
+
+    // eslint-disable-next-line
+    cy.wait(3000)
+
+    cy.get('[data-field-api-name="title"] iframe').captureIFrameAs('extension')
```
Comment (jelz) [inline test/cypress/integration/field-extension.spec.js]: Abstract or at least add todo.

File: test/cypress/integration/field-extension.spec.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    cy.setAuthTokenToLocalStorage()
+  })
+
+  it('opens first post and checks that field extension is rendered', () => {
+    cy.visit(entry(post.id))
+
+    cy.getByText(post.title).should('exist')
+
+    // eslint-disable-next-line
+    cy.wait(3000)
```
Comment (jelz) [inline test/cypress/integration/field-extension.spec.js]: can we listen to `onload` of the frame instead of hardcoding reasonable but arbitrary number?

File: test/cypress/plugins/index.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+module.exports = (on, config) => {
+  on('before:browser:launch', (browser = {}, args) => {
+    if (browser.name === 'chrome') {
+      args.push('--disable-site-isolation-trials')
```
Comment (jelz) [inline test/cypress/plugins/index.js]: Add a comment

File: test/integration/index.js [hunk 8]
```diff
@@ -1,21 +1,77 @@
...
+  await deployExtensions()
+
+  try {
+    await runCypress({
+      baseUrl: config.baseUrl
+    })
+  } catch (e) {}
+}
+
+run()
```
Comment (jelz) [inline test/integration/index.js]: Should be equivalent with no promises ;)

```js
(async function () {
  try {
    await run()
    await cleanup()
    process.exit(0)
  } catch (err) {
    console.log(err)
    await cleanup()
    process.exit(1)
  }
})()
```

File: test/integration/tasks/create-new-enviromenment.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+const printStepTitle = require('../utils').printStepTitle
+
+module.exports = async () => {
+  printStepTitle('Creating a new environment for testing')
+
+  const space = await getCurrentSpace()
+
+  const environmentId = nanoid()
+
+  await space.createEnvironmentWithId(environmentId, { name: environmentId }, 'master')
```
Comment (jelz) [inline test/integration/tasks/create-new-enviromenment.js]: Does it wait for creation to be completed? What do you get as `sys.status` right after the call is finished?

File: test/integration/utils.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+const fs = require('fs-extra')
+const os = require('os')
+const path = require('path')
+
+const rootDirectory = fs.realpathSync(process.cwd())
+const resolvePath = relativePath => path.resolve(rootDirectory, relativePath)
+
+let stepCount = 1
+
+function printStepTitle(title) {
```
Comment (jelz) [inline test/integration/utils.js]: How about `makeStepCounter` and no global `stepCount`?

Comment (suevalov) [reply test/integration/utils.js]: Could you please explain what do you have in mind?

File: test/cypress/integration/field-extension.spec.js (hunk 6 above)
Comment (suevalov) [reply test/cypress/integration/field-extension.spec.js]: Unfortunately `onload` doesn't work was expected in Cypress and it's the only working solution I could find.

File: .eslintrc.json (hunk 1 above)
Comment (suevalov) [reply .eslintrc.json]: done

File: test/cypress/integration/field-extension.spec.js (hunk 5 above)
Comment (suevalov) [reply test/cypress/integration/field-extension.spec.js]: I've abstracted the majority of the logic to `captureIFrameAs('extension')` already.
How do you see it?

File: test/integration/tasks/create-new-enviromenment.js (hunk 9 above)
Comment (suevalov) [reply test/integration/tasks/create-new-enviromenment.js]: fixed

File: docs/run-integration-tests.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
+# Run integration tests
+
+## Writing tests and running them locally
+
+1. Create `.env` file at the root of the repository with the following variables:
+
+```bash
+CONTENTFUL_CMA_TOKEN=<personal-cma-token>
```
Comment (jelz) [inline docs/run-integration-tests.md]: Link to https://www.contentful.com/r/knowledgebase/personal-access-tokens/ in a comment

File: docs/run-integration-tests.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+CONTENTFUL_SPACE=<testing-space-id>
```
Comment (jelz) [inline docs/run-integration-tests.md]: make it `CONTENTFUL_SPACE_ID`

File: docs/run-integration-tests.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+CYPRESS_BASE_URL=https://app.contentful.com
```
Comment (jelz) [inline docs/run-integration-tests.md]: I'd name it `WEB_APP_BASE_URL`

File: docs/run-integration-tests.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+TEST_LOCAL_SDK=true
```
Comment (jelz) [inline docs/run-integration-tests.md]: Add a comment saying what it means

File: docs/run-integration-tests.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+CONTENTFUL_LOCAL_TESTING_ENV=<envinonment-name-that-you-will-use-for-testing>
```
Comment (jelz) [inline docs/run-integration-tests.md]: Is it a source or a target environment?

File: test/integration/local.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
+require('dotenv').config()
+
+const buildExtensions = require('./tasks/build-extensions')
+const deployExtensions = require('./tasks/deploy-extensions')
+const createConfigurationFiles = require('./tasks/create-configuration-files')
+
+const config = {
+  cmaToken: process.env.CONTENTFUL_CMA_TOKEN,
+  spaceId: process.env.CONTENTFUL_SPACE,
+  baseUrl: process.env.CONTENTFUL_APP,
```
Comment (jelz) [inline test/integration/local.js]: This name doesn't match the readme.

File: .circleci/config.yml [hunk 17]
```diff
@@ -3,15 +3,29 @@ version: 2.1
...
     steps:
       - checkout
-      - run: npm install
+      - restore_cache:
+          keys:
+            - v2-deps-{{ .Branch }}-{{ checksum "package-lock.json" }}
+            - v2-deps-{{ .Branch }}-
+            - v2-deps-
+      - run: npm ci
+      - save_cache:
```
Comment (jelz) [inline .circleci/config.yml]: It is common for `npm` to produce unstable lock files, especially when you run it on different platforms (e.g. Mac vs Ubuntu in CI).

For this reason using checksum of the lock file for a cache key may not produce right results.

See https://github.com/contentful/doc-app/blob/master/.circleci/config.yml#L26-L38

File: lib/channel.js [hunk 18]
```diff
@@ -28,7 +28,7 @@ class Channel {
 
     this._send = createSender(sourceId, currentWindow.parent)
 
-    currentWindow.addEventListener('message', (event) => {
+    currentWindow.addEventListener('message', event => {
```
Comment (jelz) [inline lib/channel.js]: It would be nice to do prettier somewhere else. Are we 100% sure nothing was changed in the `lib/` itself (other than formatting)?

File: test/cypress/integration/field-extension.spec.js (hunk 5 above)
Comment (jelz) [reply test/cypress/integration/field-extension.spec.js]: I was thinking about something even less selector-aware, like `getExtensionForField('title').find(...).assertion()` but let's keep it as it is right now.

File: test/cypress/integration/field-extension.spec.js (hunk 6 above)
Comment (jelz) [reply test/cypress/integration/field-extension.spec.js]: Let's keep it as it is right now but I'd also research extending the Web App so it can notified us "the channel is ready"

Comment (jelz) [reply test/cypress/integration/field-extension.spec.js]: Code like that may lead to flakyiness in the future.

File: test/integration/local.js [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    'CONTENTFUL_CMA_TOKEN',
+    'CYPRESS_BASE_URL',
+    'TEST_LOCAL_SDK',
+    'CONTENTFUL_LOCAL_TESTING_ENV'
+  ].forEach(envvar => {
+    console.log(`${envvar}=${process.env[envvar]}`)
+  })
+}
+
+const run = async () => {
```
Comment (jelz) [inline test/integration/local.js]: Maybe add a comment that this one doesn't create a new environment.

Comment (jelz) [reply test/integration/local.js]: When we start doing modification it will/may become a problem.

File: docs/run-integration-tests.md (hunk 13 above)
Comment (suevalov) [reply docs/run-integration-tests.md]: `CYPRESS_BASE_URL` is a required env name for overriding `baseUrl`

File: lib/channel.js (hunk 18 above)
Comment (suevalov) [reply lib/channel.js]: Rollbacked all changes in `lib/`

File: .circleci/config.yml (hunk 17 above)
Comment (suevalov) [reply .circleci/config.yml]: Would you recommend to remove cache at all?

Comment (jelz) [reply .circleci/config.yml]: See the link - I'd revert any changes made locally in CI to `package-lock.json`
